lambdas/services/pds_api_service.py

```diff
@@ -31,32 +51,100 @@ def handle_response(self, response: Response, nhs_number: str) -> PatientDetails
+            access_token_response["issued_at"]
+        )
+        time_safety_margin_seconds = 10
+        if time() - access_token_expiration < time_safety_margin_seconds:
+            access_token = self.get_new_access_token()
 
-        pds_patient: dict = {}
+        x_request_id = str(uuid.uuid4())
 
-        for result in mock_pds_results:
-            for k, v in result.items():
-                if v == nhsNumber:
-                    pds_patient = result.copy()
+        authorization_header = {
+            "Authorization": f"Bearer {access_token}",
+            "X-Request-ID": x_request_id,
+        }
 
-        response = Response()
+        url_endpoint = endpoint + "Patient/" + nshNumber
+        pds_response = requests.get(url=url_endpoint, headers=authorization_header)
 
-        if bool(pds_patient):
-            response.status_code = 200
-            response._content = pds_patient
-        else:
-            response.status_code = 404
+        if pds_response.status_code == 401 & retry_on_expired:
+            return self.pds_request(nshNumber, retry_on_expired=False)
+        return pds_response
 
-        return response
+    def get_new_access_token(self):
+        try:
+            access_token_ssm_parameter = self.get_parameters_for_new_access_token()
+            jwt_token = self.create_jwt_token_for_new_access_token_request(
+                access_token_ssm_parameter
+            )
+            nhs_oauth_endpoint = access_token_ssm_parameter[
+                "/prs/dev/user-input/nhs-oauth-endpoint"
+            ]
+            nhs_oauth_response = self.request_new_access_token(
+                jwt_token, nhs_oauth_endpoint
+            )
+            nhs_oauth_response.raise_for_status()
+            token_access_response = nhs_oauth_response.json()
+            self.update_access_token_ssm(json.dumps(token_access_response))
+        except HTTPError as e:
+            logger.error(f"Issue while creating new access token: {e.response}")
+            raise PdsErrorException("Error accessing PDS API")
+        return token_access_response["access_token"]
+
+    def get_parameters_for_new_access_token(self):
+        parameters = [
+            "/prs/dev/user-input/nhs-oauth-endpoint",
+            "/prs/dev/user-input/pds-fhir-kid",
+            "/prs/dev/user-input/nhs-api-key",
+            "/prs/dev/user-input/pds-fhir-private-key",
+        ]
+        return self.ssm_service.get_ssm_parameters(parameters, with_decryption=True)
+
+    def update_access_token_ssm(self, parameter_value: str):
+        parameter_key = "/prs/dev-ndr/pds-fhir-access-token"
+        self.ssm_service.update_ssm_parameter(
+            parameter_key=parameter_key,
+            parameter_value=parameter_value,
+            parameter_type="SecureString",
+        )
+
+    def get_parameters_for_pds_api_request(self):
+        parameters = [
+            "/prs/dev/user-input/pds-fhir-endpoint",
```

Comment:
Lots of string paths in this file that might be better stored away as constants in a config file somewhere to avoid hard coding them?

Reply:
I've created an enum class with all of the paths, is that what you had in mind?

Reply:
Yep perfect